OpenRA.Mods.Common/UpdateRules/Rules/20221203/AbstractDocking.cs

```diff
@@ -0,0 +1,163 @@
+			{
+				var dockNode = new MiniYamlNode("DockHost", "");
+
+				dockNode.AddNode("Type", "Unload");
+
+				foreach (var value in refineyValues)
+				{
+					var node = refineryNode.ChildrenMatching(value, includeRemovals: false).FirstOrDefault();
+					if (node != null)
+					{
+						dockNode.AddNode(node);
+						refineryNode.RemoveNode(node);
+					}
+				}
+
+				if (!refineryNodes.ContainsKey(actorNode.Key) || !refineryNodes[actorNode.Key].Any(n => n.Key == "DockType"))
+					dockNode.AddNode("DockType", "Unload");
+				else
+					dockNode.AddNode(refineryNodes[actorNode.Key].First(n => n.Key == "DockType"));
+
+				var oldOffset = CVec.Zero;
+				var dockOffsetNode = refineryNode.ChildrenMatching("DockOffset", includeRemovals: false).FirstOrDefault();
+				if (dockOffsetNode != null)
+				{
+					oldOffset = dockOffsetNode.NodeValue<CVec>();
+					refineryNode.RemoveNode(dockOffsetNode);
+				}
+
+				// TODO: This conversion isn't correct
+				var newOffset = new WVec(oldOffset.X, oldOffset.Y, 0);
+				dockNode.AddNode("DockOffset", newOffset.ToString());
+
+				actorNode.AddNode(dockNode);
+			}
+
+			var harvesterNode = actorNode.ChildrenMatching("Harvester", includeRemovals: false).FirstOrDefault();
+			if (harvesterNode != null)
+			{
+				var dockClientNode = new MiniYamlNode("DockClientManager", "");
+
+				foreach (var value in harvesterValues)
+				{
+					var node = harvesterNode.ChildrenMatching(value, includeRemovals: false).FirstOrDefault();
+					if (node != null)
+					{
+						dockClientNode.AddNode(node);
+						harvesterNode.RemoveNode(node);
+					}
+				}
+
+				var nodeVoice = harvesterNode.ChildrenMatching("DeliverVoice", includeRemovals: false).FirstOrDefault();
+				if (nodeVoice != null)
+				{
+					harvesterNode.RemoveNode(nodeVoice);
+					nodeVoice.RenameKey("Voice");
+					dockClientNode.AddNode(nodeVoice);
+				}
+
+				var nodeColor = harvesterNode.ChildrenMatching("DeliverLineColor", includeRemovals: false).FirstOrDefault();
+				if (nodeColor != null)
+				{
+					harvesterNode.RemoveNode(nodeColor);
+					nodeColor.RenameKey("DockLineColor");
+					dockClientNode.AddNode(nodeColor);
+				}
+
+				var nodeQueue = harvesterNode.ChildrenMatching("UnloadQueueCostModifier", includeRemovals: false).FirstOrDefault();
+				if (nodeQueue != null)
+				{
+					harvesterNode.RemoveNode(nodeQueue);
+					nodeQueue.RenameKey("OccupancyCostModifier");
+					dockClientNode.AddNode(nodeQueue);
+				}
+
+				var nodeDelay = harvesterNode.ChildrenMatching("SearchForDeliveryBuildingDelay", includeRemovals: false).FirstOrDefault();
+				if (nodeDelay != null)
+				{
+					harvesterNode.RemoveNode(nodeDelay);
+					nodeDelay.RenameKey("SearchForDockDelay");
+					dockClientNode.AddNode(nodeDelay);
+				}
+
+				harvesterNode.RenameChildrenMatching("DeliveryBuildings", "DockType", includeRemovals: false);
+				harvesterNode.RemoveNodes("MaxUnloadQueue", includeRemovals: false);
+
+				actorNode.AddNode(dockClientNode);
```

Comment:
You could make this section cleaner and easier to maintain by using a single local function for the YAML retrieval:

```C#
MiniYamlNode GetFirstMatchingChild(MiniYamlNode node, string matchingString) => node.ChildrenMatching(matchingString, includeRemovals: false).FirstOrDefault();
```

then calling it for each instance of `actorNode.ChildrenMatching()` and each `harvestorNode.ChildrenMatching()` as per below:

```C#
var harvesterNode = GetFirstMatchingChild(actorNode, "Harvester");
```

and 


```C#
var node = GetFirstMatchingChild(harvestorNode, value);
```

etc. I have tested this and the above syntax works.

Reply:
Changed it in another way. But I dunno if it was even worth the effort, update rules needed to be optimised or maintained